### PR TITLE
Chat operations as Celery tasks

### PR DESCRIPTION
**Pull request by rgossiaux**
Figured I might as well do this now. The discord API is fairly
responsive but I'd noticed it was still introducing some lag to a lot of
interactions.

### CODE REVIEW

File: chat/models.py [hunk 1]
```diff
@@ -116,11 +116,11 @@ def send_message(self, msg):
             service = self.get_service()
             service.send_message(self.text_channel_id, msg)
 
-    def handle_tag_added(self, puzzle, tag):
-        self.get_service().handle_tag_added(self, puzzle, tag)
+    def handle_tag_added(self, puzzle, tag_name):
+        self.get_service().handle_tag_added(puzzle, tag_name)
```
Comment (rawxfish) [inline chat/models.py:120]: What's the reason for having handle_tag_added jump around from tasks to the chat model back to discord_chat_service.py?

Comment (asdfryan) [reply chat/models.py:120]: IMO our initial plan was not to expose the service on any views (or tasks). This is WAI IMO.

File: chat/tasks.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    try:
+        puzzle.chat_room.unarchive_channels()
+    except Exception as e:
+        logger.warn(f"Chat operations failed with error: {e}")
+
+
+@shared_task
+def handle_tag_added(puzzle_id, tag_name):
+    puzzle = Puzzle.objects.get(id=puzzle_id)
+    puzzle.chat_room.handle_tag_added(puzzle, tag_name)
```
Comment (asdfryan) [inline chat/tasks.py:48]: Should we also wrap this with try except? Same for removed.

**Review by asdfryan — APPROVED** [review]
Nice! Approving with 1 (somewhat unrelated) nit. I'd like to get this in ASAP so I can iterate on my discord issues on top of it.

File: chat/models.py (hunk 1 above)
Comment (rawxfish) [reply chat/models.py:120]: Gotcha, makes sense

**Review by asdfryan — CHANGES_REQUESTED** [review]
Looks like some tests are failing.

File: api/views.py [hunk 3]
```diff
@@ -224,16 +222,11 @@ def create(self, request, **kwargs):
...
-                    chat_room.get_service().announce(msg)
-                    chat_room.send_message(msg)
-                except Exception as e:
-                    logger.warn(f"Chat operations failed with error: {e}")
+                transaction.on_commit(
+                    lambda: chat.tasks.create_chat_for_puzzle.delay(puzzle.id)
+                )
             else:
                 logger.warn("Chat room not created for puzzle %s" % name)
-                chat_room = None
```
Comment (asdfryan) [inline api/views.py:236]: I think you have to keep this line. Otherwise line 231 will fail.

**Comment by rgossiaux** [comment]
I'm working for the next 4-5 hours--happy to merge after that, but if you
need this in sooner to unblock would you mind fixing your nit and the test
(sorry about that) and merging it? I agree with your suggestions.

**Review by asdfryan — APPROVED** [review]
Nits/tests are now fixed. Merging this. :)

**Comment by rgossiaux** [comment]
Thank you!!

**Pull request merged**